test/distributed/test_hooks.py

```diff
@@ -138,6 +141,65 @@ def check_op(idx, coll_name):
         check_op(0, "ALLGATHER")
         check_op(1, "ALLREDUCE")
 
+    def _test_failure(self):
+        # it's ok to access them directly since there's a single bg thread poking at them.
+        starts = []
+        ends = []
+        cv = threading.Condition()
+
+        def coll_start(status):
+            starts.append(status)
+            print(f"col_start {len(starts)} rank{self.rank}")
+
+        def coll_end(status):
+            ends.append(status)
+            print(f"col_end {len(ends)} rank{self.rank}")
+            if len(ends) == 2:
+                with cv:
+                    cv.notify()
+
+        dhooks.register_collective_start_hook(coll_start)
+        dhooks.register_collective_end_hook(coll_end)
+
+        tensor = torch.ones([2, 3]).cuda(self.rank) * self.rank
+        tensor_list = [torch.empty_like(tensor) for _ in range(self.world_size)]
+
+        dist.all_gather(tensor_list, tensor)
+
+        tensor2 = torch.ones([2, 3]).cuda(self.rank) * self.rank
+        if self.rank == 2:
```

Comment:
are we assuming world_size == 2? is there a testing macro we can wrap to ensure that if so?

Reply:
the test harness hardcode world size 4